server/database.js

```diff
@@ -10,87 +10,91 @@ import { clearTask, scheduleTask } from './tasks.js'
 
 // $garbage will be the parent of deleted objects
 global.$garbage = new DatabaseObject()
 
 export function createObject(parent, data) {
-    if (parent.constructor.name !== 'DatabaseObject')
+    if (!(parent instanceof DatabaseObject))
         throw new Error('Parent must be a database object')
     const obj = Object.create(parent)
     Object.assign(obj, data)
     obj.id = ulid() // Created objects will always get a new ID as an anti-footgun measure
+    obj.parent = parent
+    obj.children = new Set()
     db[obj.id] = obj
+    parent.children.add(obj)
     return obj
 }
 
 export function deleteObject(obj) {
     if (obj.constructor.name !== 'DatabaseObject')
         throw new Error('Object is not a database object')
-    if (children(obj).length)
+    if (obj.children.length)
         throw new Error('Cannot delete object with children')
     // Obj might still be referenced by other objects.  We change its parent to $garbage so it's obvious that
     // the object is dead when it's referenced
     changeParent(obj, $garbage)
     delete db[obj.id]
 }
 
-export function changeParent(obj, parent) {
-    if (parent.constructor.name !== 'DatabaseObject')
+export function changeParent(obj, newParent) {
+    if (!(newParent instanceof DatabaseObject))
         throw new TypeError('Invalid parent object')
     // Make sure the new parent isn't a descendant of obj
-    if (descendants(obj).includes(parent))
+    if (descendants(obj).includes(newParent))
         throw new Error('New parent is a descendant of the object')
-    Object.setPrototypeOf(obj, parent)
-}
-
-export function children(parent) {
-    const kids = []
-    for (const obj of Object.values(db)) {
-        if (Object.getPrototypeOf(obj) === parent)
-            kids.push(obj)
-    }
-    return kids
-}
-
-export function ancestors(obj) {
-    const ancs = []
-    let next = parent(obj)
-    while (next) {
-        ancs.push(next)
-        next = parent(next)
-    }
-    return ancs
-}
-
-export function descendants(obj) {
-    const descs = []
-    for (const kid of children(obj)) {
-        descs.push(kid, ...descendants(kid))
-    }
-    return descs
-}
-
-export function leaves(obj) {
-    const lvs = []
-    for (const kid of children(obj)) {
-        const d = descendants(kid)
-        if (d.length)
-            lvs.push(...leaves(kid))
-        else
-            lvs.push(kid)
-    }
-    return lvs
+    const oldParent = obj.parent
+    if (!oldParent)
+        throw new Error('Object has no parent.  Are you trying to reparent the root object?')
+    oldParent.children.delete(obj)
+    Object.setPrototypeOf(obj, newParent)
+    obj.parent = newParent
+    newParent.children.add(obj)
 }
 
-export function parent(obj) {
-    if (obj.constructor.name !== 'DatabaseObject')
+// This function will not be used unless we think something terrible has happened.  Parent will normally be looked
+// up via the .parent property on DB objects.
+export function findParent(obj) {
```

Comment:
Do you even still need it to exist?

Reply:
No!  I meant to remove it, and now I have.  Thanks for calling that out.